src/Moq/ExpressionExtensions.cs

```diff
@@ -362,10 +362,15 @@ internal static PropertyInfo GetReboundProperty(this MemberExpression expression
 			// we "upgrade" to the derived property.
 			if (property.DeclaringType != expression.Expression.Type)
 			{
+				var parameterTypes = new ParameterTypes(property.GetIndexParameters());
 				var derivedProperty = expression.Expression.Type
 					.GetMember(property.Name, MemberTypes.Property, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance)
 					.Cast<PropertyInfo>()
-					.SingleOrDefault(p => p.PropertyType == property.PropertyType);
+					.SingleOrDefault(p =>
+					{
+						return p.PropertyType == property.PropertyType
+							&& new ParameterTypes(p.GetIndexParameters()).CompareTo(parameterTypes, true, false);
+					});
```

Comment:
This could possibly be optimized further. Ideally, we'd avoid the calls to `.GetIndexParameters()`, as well as the comparison, by checking first whether both `PropertyInfo`s are of the same type (property vs. indexer), and only doing the comparison when both are indexers.